### DISCUSSION TITLE
sql server implementation?

### DISCUSSION THREAD

**Discussion by himadrinath** [ai/nanoid#412] [General]
is their any implementation available for ms sql server?

**Comment by ai** [comment]
I don’t know about such implementation

**Comment by himadrinath** [comment]
i asked chatgpt

here is the response

Nano ID is a small, secure, URL-friendly, unique string ID generator library. It generates random IDs that contain 21 URL-friendly characters (A-Za-z0-9_~) in length.

Here's an example of how to create a Nano ID in Microsoft SQL Server using T-SQL:

```
CREATE FUNCTION dbo.GenerateNanoId ()
RETURNS NVARCHAR(21)
AS
BEGIN
    DECLARE @characters nvarchar(50) = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_~';
    DECLARE @id nvarchar(21) = '';
    DECLARE @len int = LEN(@characters);

    WHILE LEN(@id) < 21
    BEGIN
        SET @id = @id + SUBSTRING(@characters, CAST((RAND() * @len) AS INT) + 1, 1);
    END

    RETURN @id;
END;
```
Explanation:

1.  The CREATE FUNCTION statement creates a user-defined function named GenerateNanoId that returns an NVARCHAR string with a length of 21.
2. The @characters variable holds the set of characters that will be used to generate the Nano ID.
3. The @id variable holds the generated Nano ID.
4. The @len variable holds the length of the @characters variable.
5. The WHILE loop continues generating characters and appending them to the @id variable until it reaches the desired length of 21.
6. The SUBSTRING function is used to extract a single character from the @characters string at a random index, and the RAND() function is used to generate the random index.
7. The final line RETURN statement returns the generated Nano ID.

This function can be used within a stored procedure or function to generate a Nano ID for each row or record inserted into a table. It uses a set of characters and a loop to generate a random string of 21 characters in length.

**Reply by himadrinath** [reply]
after running this query

`Invalid use of a side-effecting operator 'rand' within a function.`

and i also don't think that logic is not completely right to implement the nanoid

**Reply by ai** [reply]
Check your SQL server docs that `RND()` is using hardware randomness.

Without hardware randomness there will be to big chances of having ID collision on different machines.
